clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/LambdaCachingAgent.java

```diff
@@ -212,4 +248,152 @@ public CacheResult loadData(ProviderCache providerCache) {
+    }
+
+    Map<String, Object> attributes = new HashMap<String, Object>();
+    attributes.put("name", appName);
+    Map<String, Collection<String>> evictions = Collections.emptyMap();
+
+    Collection<String> existingFunctionRel = null;
+
+    CacheData application = providerCache.get(APPLICATIONS.ns, appKey);
+
+    if (null != application && null != application.getRelationships()) {
+      existingFunctionRel = application.getRelationships().get(LAMBDA_FUNCTIONS.ns);
+    }
+
+    Map<String, Collection<String>> relationships = new HashMap<String, Collection<String>>();
+
+    if (null != existingFunctionRel && !existingFunctionRel.isEmpty()) {
+      if (null == functionResult && existingFunctionRel.contains(functionKey)) {
+        existingFunctionRel.remove(functionKey);
+        evictions.put(LAMBDA_FUNCTIONS.ns, Collections.singletonList(functionKey));
+      } else {
+        existingFunctionRel.add(functionKey);
+      }
+
+    } else {
+      existingFunctionRel = Collections.singletonList(functionKey);
+    }
+
+    relationships.put(LAMBDA_FUNCTIONS.ns, existingFunctionRel);
+    DefaultCacheData cacheData = new DefaultCacheData(appKey, attributes, relationships);
+    DefaultCacheResult defaultCacheresults =
+        new DefaultCacheResult(
+            Collections.singletonMap(
+                com.netflix.spinnaker.clouddriver.core.provider.agent.Namespace.APPLICATIONS.ns,
+                Collections.singletonList(cacheData)));
+
+    return new OnDemandAgent.OnDemandResult(getAgentType(), defaultCacheresults, evictions);
+  }
+
+  @Override
+  public Collection<Map> pendingOnDemandRequests(ProviderCache providerCache) {
+    return null;
+  }
+
+  @Override
+  public String getOnDemandAgentType() {
+    return getAgentType() + "-OnDemand";
+  }
+
+  @Override
+  public OnDemandMetricsSupport getMetricsSupport() {
+    return metricsSupport;
+  }
+
+  private Boolean validKeys(Map<String, ? extends Object> data) {
+    return (data.containsKey("functionName")
+        && data.containsKey("credentials")
+        && data.containsKey("region"));
+  }
+
+  protected String combineAppDetail(String appName, String functionName) {
+    Names functionAppName = Names.parseName(functionName);
+    if (null != functionAppName) {
+      return functionAppName.getApp().equals(appName)
+          ? functionName
+          : (appName + "-" + functionName);
+    } else {
+      throw new IllegalArgumentException(
+          String.format("Function name {%s} contains invlaid charachetrs ", functionName));
+    }
+  }
+
+  private List<String> getTargetGroupNames(AWSLambda lambda, String functionName) {
+    List<String> targetGroupNames = new ArrayList<>();
+    try {
+      GetPolicyResult result =
+          lambda.getPolicy(new GetPolicyRequest().withFunctionName(functionName));
+      String json = result.getPolicy();
+      Policy policy = Policy.fromJson(json);
+      for (Statement statement : policy.getStatements()) {
```

Comment:
as I commented on the previous incarnation of this PR - I'd like to see this flattened out via the streams api

Reply:
Replaced this with Streams API.